Add SERVER_ENV_CONFIG to configure vars from env. variable

guewen opened the pull request:
It allows to inject secret values without storing them in files, change values outside of the "project distribution" (docker, git, ...).

Commit:
Add SERVER_ENV_CONFIG to configure vars from env. variable

elicoidal commented on server_environment/README.rst:
```diff
@@ -47,8 +47,12 @@ You should then edit the settings you need in the
+* additional configuration can be passed in the environment variable
+  ``SERVER_ENV_CONFIG``, overriding any values set in the configuration files.
+  This is a multi-line environment variable in the same configparser format than
+  the files
```
add period

guewen replied:
added

elicoidal reviewed (approved):
LGTM

sbidoul commented on server_environment/serv_config.py:
```diff
@@ -97,9 +97,20 @@ def _load_config():
+        except configparser.Error as err:
+            raise Exception(
+                'SERVER_ENV_CONFIG content could not be parsed: %s'
+                % (err,)
```
Have you checked `err` does not reveal the content of the of the environment variable?

guewen replied:
Good catch, it does!

```
2018-07-19 06:18:57,343 1 CRITICAL odoodb odoo.modules.module: Couldn't load module server_environment
2018-07-19 06:18:57,343 1 CRITICAL odoodb odoo.modules.module: SERVER_ENV_CONFIG content could not be parsed: File contains no section headers.
file: '<string>', line: 1
'my super secret but invalid format'
2018-07-19 06:18:57,345 1 WARNING odoodb odoo.modules.loading: Transient module states were reset
2018-07-19 06:18:57,345 1 ERROR odoodb odoo.modules.registry: Failed to load registry
Traceback (most recent call last):
  File "/opt/odoo/external-src/server-env/server_environment/serv_config.py", line 107, in _load_config
    config_p.read_string(env_config)
  File "/usr/lib/python3.5/configparser.py", line 720, in read_string
    self.read_file(sfile, source)
  File "/usr/lib/python3.5/configparser.py", line 715, in read_file
    self._read(f, source)
  File "/usr/lib/python3.5/configparser.py", line 1077, in _read
    raise MissingSectionHeaderError(fpname, lineno, line)
configparser.MissingSectionHeaderError: File contains no section headers.
file: '<string>', line: 1
'my super secret but invalid format'
```

I'll see how to prevent that.

sbidoul commented:
Excellent!

guewen commented:
@sbidoul I tested a range of potential errors and the config printed is minimal, the most sensitive one is the first below, if you miss a section header, it shows the first line. We could show only the class name in the message (`<class configparser.DuplicateSectionError>`, `<class configparser.MissingSectionHeaderError>`, ...), should I do that?

```
Exception: SERVER_ENV_CONFIG content could not be parsed: File contains no section headers.
file: '<string>', line: 1
'sftp_password=password\n'
```


```
Exception: SERVER_ENV_CONFIG content could not be parsed: Source contains parsing errors: '<string>'
        [line  2]: 'sftp_port\n'
```

```
Exception: SERVER_ENV_CONFIG content could not be parsed: While reading from '<string>' [line  3]: option 'sftp_port' in section 'storage_backend.SFTP Backend' already exists
```

```
Exception: SERVER_ENV_CONFIG content could not be parsed: While reading from '<string>' [line  6]: section 'ir_mail_server' already exists
```

```
configparser.InterpolationSyntaxError: bad interpolation variable reference '%(test)'
```

```
configparser.InterpolationMissingOptionError: Bad value substitution: option 'type' in section 'incoming_mail.foo' contains an interpolation key 'test' which is not a valid option name. Raw value: 'imap%(test)s'
```

sbidoul commented:
It's good. I think I raised that more out of curiosity than anything else, and I would not have thought about it if you had not written that try/except. If something had to be done that would be in ConfigParser I guess.

Commit:
Add SERVER_ENV_CONFIG_SECRET alongside SERVER_ENV_CONFIG

Allows to isolate the secrets in your deployment

TDu commented on server_environment/README.rst:
```diff
@@ -47,8 +47,12 @@ You should then edit the settings you need in the
-* finally, you can override or extend values in the main configuration
-  file of you instance.
+* you can override or extend values in the main configuration
+  file of you instance;
```
missing 'r' : your instance ?

guewen replied:
fixed

TDu reviewed (approved):
:+1: 
Just a small typo maybe.

guewen commented:
@sbidoul @max3903 @TDu @elicoidal @vrenaville 

Thanks for your reviews. While setting up the deploy using this, I observed that we split our configuration options in 2 parts: most of the configuration for a service is in plain-text - because it's convenient to see them in clear - and the rest of the configuration is secrets which are encrypted. For this reason, I added a `SERVER_ENV_CONFIG_SECRET`, which is handled the same way than `SERVER_ENV_CONFIG`, it's only a convenience for the deploy.
 I hope you won't mind my additional commit.

elicoidal commented on server_environment/serv_config.py:
```diff
@@ -97,9 +99,21 @@ def _load_config():
     config_p.read(system_base_config.rcfile)
     config_p.remove_section('options')
 
+    for varname in ENV_VAR_NAMES:
```
@guewen When reading the code, I am not sure what is the value of splitting.
You do not log any of them anymore so no risk of showing secret content.
I would understand that the main advantage is to make public or default one while the other is set up in a specific file in a server.
If that is your view, I would suggest to add an explanation in the README about how to use both variables

guewen replied:
> I would understand that the main advantage is to make public or default one while the other is set up in a specific file in a server.

Yes, specifically we have something like this in our deployments:

``rancher.public.env`` is a file which contains all the public environment variables in clear text:

```
export WORKERS='8'
export MAX_CRON_THREADS='1'
export LOG_LEVEL=info
export LOG_HANDLER=":INFO"
export DB_MAXCONN=5

# server environment
export SERVER_ENV_CONFIG="
[storage_backend.my-sftp]
sftp_server=10.10.10.10
sftp_login=foo
sftp_port=22200
directory_path=Odoo
"
```

``rancher.env.gpg`` which is encrypted and contains secrets:

```
export DB_PASSWORD='xxxxxxxxx'
export KEYCHAIN_KEY='egS_Oxxxxxxxxxxxxxxxxxxxxxxxxxxx='
export SERVER_ENV_CONFIG_SECRET="
[storage_backend.my-sftp]
sftp_password=xxxxxxxxx
"
```

I think this is a common pattern in deployment, furthermore with Docker now. If we wanted to support only the `SERVER_ENV_CONFIG`, we would have to merge the content of the the 2 files beforehand which would be a pain...

I'll add a few lines more in the documentation.

elicoidal replied:
:+1: We have similar pattern too.

guewen replied:
@elicoidal does my last commit improves the understanding?

elicoidal replied:
Perfect!

Commit:
Improve documentation regarding variables
